eip/EIPS/eip-4337.md

```diff
@@ -271,23 +280,39 @@ In order to distinguish between them, there is a call to the NUMBER opcode (`blo
-5. The third (validateUserOpSignature) is the first view call after the "NUMBER" marker. If that aggregator is banned, then this staticcall MUST revert immediately, as if it was called with zero gas.
-6. Any `CALL` or `CALLCODE` during validation has `value=0`, except for the transfer from the wallet to the entry point.
-7. No `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` results in an out-of-gas revert.
-8. No `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` to addresses with `EXTCODESIZE=0`.
-9. Any `GAS` opcode is followed immediately by one of { `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` }.
-10. `EXTCODEHASH` of every address accessed (by any opcode) does not change between first and second simulations of the op.
-11. If `op.initcode.length != 0` , allow only one `CREATE2` opcode call (in the validateUserOp block), otherwise forbid `CREATE2`.
-
-Since the wallet is allowed to access its own entry point deposit in order to top it up when needed, the client must know the storage slot in order to whitelist it. The entry point therefore implements the following view function:
-
-```c++
-function getSenderStorage(address sender) external view returns (uint256[] memory senderStorageCells)
-```
+2. The first (validateUserOp) call is allowed to access storage slot of the wallet itself (see Storage access by Slots, below), 
+   but only on contracts that are not part of the current bundle (that is, not in a paymaster or any other wallet)
+   Note that the wallet is allowed (and expected) to deposit to its own balance in the EntryPoint. That is allowed by this rule.
+3. The second (validatePaymasterUserOp), the paymaster may access storage slots of itself AND of the wallet, above.
+4. The third (validateUserOpSignature) view-call, the aggregator is allowed to access storage slots of itself and of the wallet. 
+5. Any `CALL` or `CALLCODE` during validation has `value=0`, except for the transfer from the wallet to the entry point.
+6. No `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` results in an out-of-gas revert.
+7. No referenced address (`CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL`, `EXTCODESIZE`, `EXTCODEHASH` to addresses with `EXTCODESIZE=0`.
+8. Any `GAS` opcode is followed immediately by one of { `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` }.
+9. `EXTCODEHASH` of every address accessed (by any opcode) does not change between first and second simulations of the op.
+10. If `op.initcode.length != 0` , allow only one `CREATE2` opcode call (in the validateUserOp block), otherwise forbid `CREATE2`.
+
+#### Storage access by slots
+
+During validation, a contract `A` may reference storage of
+1. Slots of contract `A` address itself. 
+2. Slot `A` on any other address. 
+3. Slots of type `keccak256(A || X)` on any other address. (to cover `mapping(address => value)`, which is usually used for balance in ERC20 tokens). 
```

Comment:
How does the bundler know the slot is accessed this way? Is there a specific pattern of opcodes to detect? If so, do we become reliant on Solidity implementation details?

Reply:
It does rely on the solidity pattern of using keccak to handle mapping, although it could be generalized a bit to whitelist (recursively) any keccak result that included `A` (or a previously whitelisted address) in its input, so it would cover `keccak(A||X)` but also `keccak(X||A||Const)` or whatever.

Keep in mind that these rules are implemented by bundlers, not enforced on-chain by EntryPoint.  The community can always add more methods, so when EVM gets a new precompile to support `shinyNewHash` and then the `newShinyLanguage` compiler determines slots address using `newShinyHash(🌞||A||X)`, we just add it to the specs and bundlers can add it to their implementation.

Reply:
these 2 keccak rules are the way currently solidity (and I think vyper?) use mappings. 
in theory, we should block any `keccak( * || A || *)` and `keccak( * || OWN || * )`.
These are harder to implement, and we'll add them in case there is a language that actually uses a more complex reference-address-inside-a-keccak format.

Reply:
That's fine, but I'm curious - why is it harder to implement?  We know the memory range which keccak is going to consume.  Why can't we mark the entire keccak operation as "safe" if any word-aligned offset contains `A` or `OWN`, and then whitelist the result?  Is this implementation really more complicated?

Reply:
yes it is possible. scanning the memory block (can be as you suggest, in word-aligned jumps), and each slot verify against the hashset we keep of valid values (A, and collected "OWN" values) it is a loop for each keccak, but a small one (currently the block size is 64 (2 words). can be limit it to <5 words. each extra word is needed to define a new hash dimension, and EVM does not support dynamic languages with run-time dimensions...